From dc5c01e168666e5ff9f05592737e4ceef3138aa7 Mon Sep 17 00:00:00 2001
From: Amir Vadai <avadai@redhat.com>
Date: Thu, 11 Sep 2014 12:00:48 -0400
Subject: [ethernet] mlx4: User prio mapping gets corrupted when changing number of channels

Message-id: <2cc0ed03bcbec58342bac77290ed0396fe70809e.1410425016.git.avadai@redhat.com>
Patchwork-id: 93696
O-Subject: [Patch RHEL7.1 V1 112/197] net/mlx4_en: User prio mapping gets corrupted when changing number of channels
Bugzilla: 1107617 1107618 1107619 1107620
RH-Acked-by: Jes Sorensen <Jes.Sorensen@redhat.com>

From: Ido Shamay <idos@mellanox.com>

When using ethtool set_channels, mlx4_en_setup_tc is always called, even
when it was not configured. Fixed code to call mlx4_en_setup_tc() only
if needed.

Signed-off-by: Ido Shamay <idos@mellanox.com>
Signed-off-by: Amir Vadai <amirv@mellanox.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
(cherry picked from commit f5b6345ba8daf65a96ba4931c63624936908cf0f)
Signed-off-by: Amir Vadai <avadai@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/mellanox/mlx4/en_ethtool.c b/drivers/net/ethernet/mellanox/mlx4/en_ethtool.c
index c373604..a72d99f 100644
--- a/drivers/net/ethernet/mellanox/mlx4/en_ethtool.c
+++ b/drivers/net/ethernet/mellanox/mlx4/en_ethtool.c
@@ -1151,7 +1151,8 @@ static int mlx4_en_set_channels(struct net_device *dev,
  netif_set_real_num_tx_queues(dev, priv->tx_ring_num);
  netif_set_real_num_rx_queues(dev, priv->rx_ring_num);
 
- mlx4_en_setup_tc(dev, MLX4_EN_NUM_UP);
+ if (dev->num_tc)
+  mlx4_en_setup_tc(dev, MLX4_EN_NUM_UP);
 
  en_warn(priv, "Using %d TX rings\n", priv->tx_ring_num);
  en_warn(priv, "Using %d RX rings\n", priv->rx_ring_num);
-- 
1.7.1